### PR TITLE
Build docker images for multiple architectures

### PR DESCRIPTION
**Pull request by pestotoast**
Related to #28
Instead of manually building the images, this github action will build all the docker images on every push to the main branch and publish them to dockerhub. 

- Only pushes into the **development** tag (not the latest)
- Build the images for multiple architechtures: **amd64** and **arm64**
- Some builds for arm64 images don't work yet. I also have not tested on arm64 yet.
- Needs DOCKERHUB_USERNAME and DOCKERHUB_TOKEN configured as secret
- Here's how the jobs look like: https://github.com/pestotoast/all-in-one/actions/runs/1536966830
- Here's how it looks when pushed to dockerhub: https://hub.docker.com/u/agebhart

### CODE REVIEW

**Comment by szaimen** [comment]
Hi, thanks for you PR!
Do you think it is possible to build multiarch images? I think this is what we are heading towards with https://github.com/nextcloud/all-in-one/issues/28. Feel free to click on the links that you find there :)

**Comment by pestotoast** [comment]
> Do you think it is possible to build multiarch images?

Yes, see [e895aaff](
https://github.com/pestotoast/all-in-one/commit/e895aaffbb490a8b09d86863c004fefe0943cee6#diff-87ee5504a3e25ac558b343724c905f2f7949e8cec3d92b9c4300bb922afa164f) for example.

I thought it's better to keep the images in seperate tags as long as some of arm64 builds still fail. If the arm64 build fails the amd64 image does not get pushed to dockerhub.

**Comment by szaimen** [comment]
> I thought it's better to keep the images in seperate tags as long as some of arm64 builds still fail. If the arm64 build fails the amd64 image does not get pushed to dockerhub.

makes probably sense, yes

> Yes, see [e895aaff](https://github.com/pestotoast/all-in-one/commit/e895aaffbb490a8b09d86863c004fefe0943cee6#diff-87ee5504a3e25ac558b343724c905f2f7949e8cec3d92b9c4300bb922afa164f) for example.

Is there already a mulltiarch image that builds correctly in your repo? We will need it to investiage how the manifest at the docker hub registry looks like as we are using that as update check.

**Comment by szaimen** [comment]
Here is an example for the shell:
`export TOKEN=$(curl "https://auth.docker.io/token?service=registry.docker.io&scope=repository:nextcloud/all-in-one:pull" | jq -r .token)`
`curl --head  -H "Authorization: Bearer $TOKEN" https://registry-1.docker.io/v2/nextcloud/all-in-one/manifests/develop -H 'Accept: application/vnd.docker.distribution.manifest.v2+json'`

**Comment by pestotoast** [comment]
Okay I tested your script for my `aio-talk` image (mastercontainer does not build successfully yet). 
Here's how the `aio-talk `image looks on dockerhub: https://hub.docker.com/r/agebhart/aio-talk/tags The multiarch image is in the `development` tag.


```
ag@kronos ~
$ export TOKEN=$(curl "https://auth.docker.io/token?service=registry.docker.io&scope=repository:agebhart/aio-talk:pull" | jq -r .token)
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  4371    0  4371    0     0   9664      0 --:--:-- --:--:-- --:--:--  9670
ag@kronos ~
$ curl --head -H "Authorization: Bearer $TOKEN" https://registry-1.docker.io/v2/agebhart/aio-talk/manifests/development -H 'Accept: application/vnd.docker.distribution.manifest.v2+json'
HTTP/1.1 200 OK
content-length: 2614
content-type: application/vnd.docker.distribution.manifest.v2+json
docker-content-digest: sha256:46d5eae25acdcd4f73f6fb220096d342f9cf631d52a6014e49a6df5fdf1d875f
docker-distribution-api-version: registry/2.0
etag: "sha256:46d5eae25acdcd4f73f6fb220096d342f9cf631d52a6014e49a6df5fdf1d875f"
date: Sat, 04 Dec 2021 16:13:12 GMT
strict-transport-security: max-age=31536000
ratelimit-limit: 100;w=21600
ratelimit-remaining: 82;w=21600
```

The `docker-content-digest` is the one for the amd64 version on dockerhub:
grafik
I'm not sure how we can get the checksum for the arm64 version though.

**Comment by szaimen** [comment]
I'll think we need to investigate that before going foward. But there must be a way. Maybe by sending an additional header or so...

**Comment by szaimen** [comment]
cc @LukasReschke as he is the Pro in finding such things out ;)

**Comment by szaimen** [comment]
looks like it is possible that multiarch manifests are actually not supported by the docker registry api:
image
(from https://docs.docker.com/registry/spec/api/)

But lets see what @LukasReschke says :)

**Comment by szaimen** [comment]
Okay, after thinking about this a bit more I think it is fine to have different tags for arm64 as we don't have to rework the updatecheck logic or actually no logic at all. The only small downsize is that people with arm64 will need to use a different tag but I guess that should be fine. Also being able to slowly fix the building problems is a big plus.

@LukasReschke what do you think about adding DOCKERHUB_USERNAME  and DOCKERHUB_TOKEN to the secrets? Also do you mind disabling the autobuilds on docker hub when we merge this? ;)

File: .github/workflows/build_images.yml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
+      
+      - name: Docker metadata
+        id: metadata
+        uses: docker/metadata-action@v3
+        with:
+          images: |
+            ${{ secrets.DOCKERHUB_USERNAME }}/${{ matrix.container.name }}
```
Comment (szaimen) [inline .github/workflows/build_images.yml:50]: I think it should be
```suggestion
            nextcloud/${{ matrix.container.name }}
```

**Comment by LukasReschke** [comment]
As the Nextcloud organization membership is really broad I am a bit reluctant to add publishing secrets to this repository. Especially as I don't see a way to lockdown publishes on DockerHub just to a given tag. (e.g. `develop` and not `latest`) 

Even if we can lockdown the secret inside GitHub to a given branch using environments, I would still feel a bit uncomfortable.

I wonder if we can use Docker Content Trust here? If we had a way to sign the `latest` tag and enforce that on pulling, that would be quite nice. - https://docs.docker.com/engine/security/trust/

**Comment by szaimen** [comment]
> Especially as I don't see a way to lockdown publishes on DockerHub just to a given tag. (e.g. `develop` and not `latest`)

Isn't that already done with https://github.com/nextcloud/all-in-one/pull/42/files#diff-afa3d7d08b5e9c094aec36d7fe957d9bdb65501dc49153cac56eacdf4ca1c55dR15-R16 ? :)

**Comment by LukasReschke** [comment]
> Isn't that already done with https://github.com/nextcloud/all-in-one/pull/42/files#diff-afa3d7d08b5e9c094aec36d7fe957d9bdb65501dc49153cac56eacdf4ca1c55dR15-R16 ? :)

But that is only a limitation for the action, not the secret. Right? 

e.g. as a malicious person I could still push a malicious branch to extract the secret, or as a malicious admin modify the action.

**Comment by szaimen** [comment]
> I wonder if we can use Docker Content Trust here? If we had a way to sign the `latest` tag and enforce that on pulling, that would be quite nice. - https://docs.docker.com/engine/security/trust/

Sounds interesting. I guess its worth investigating 👍

**Comment by szaimen** [comment]
> or as a malicious admin modify the action.

I guess that would be possible, yes.

**Comment by szaimen** [comment]
> > I wonder if we can use Docker Content Trust here? If we had a way to sign the `latest` tag and enforce that on pulling, that would be quite nice. - https://docs.docker.com/engine/security/trust/
> 
> Sounds interesting. I guess its worth investigating 👍

What would you recommend? only signing the `latest` releases or the develop releases as well?
Signing could than be added to the script that we use to publish from develop to latest, no?

**Comment by LukasReschke** [comment]
> What would you recommend? only signing the latest releases or the develop releases as well?
> Signing could in the first case be added to the script that we use to publish from develop to latest, no?

Only sign the `latest` tag. We keep `develop` unsigned and will have the DockerHub secrets inside a GitHub environment.

Some open points here:

1. If someone gets access to the GitHub environment can they get a new signing key or delete the old one? I assume not but am not sure :)
2. There probably needs to be changes to make the management container only download signed containers for `latest`. Maybe that is a Docker config somewhere?

**Comment by szaimen** [comment]
> * If someone gets access to the GitHub environment can they get a new signing key or delete the old one? I assume not but am not sure :)

We don't safe the signing keys on github, no?

> * There probably needs to be changes to make the management container only download signed containers for `latest`. Maybe that is a Docker config somewhere?

Based on https://docs.docker.com/engine/security/trust/#client-enforcement-with-docker-content-trust we need to enforce it on client side but not sure how to do that only by using the api...

**Comment by szaimen** [comment]
@LukasReschke Can we maybe restrict publishing secrets to be only able to publish to a specific tag? This would probably already solve most concerns, no?

**Comment by szaimen** [comment]
@LukasReschke I have another idea: what about actually adding an autobuild progress to docker hub that also automatically builds the arm64 image and pushes it to the develop-arm64 tag automatically? Would that be possible?

**Comment by LukasReschke** [comment]
Conclusion after discussion with @szaimen will be that we can use GitHub Actions but will build it in the nextcloud-releases org (which will mirror this repository)

File: .github/workflows/build_images.yml (hunk 1 above)
Comment (szaimen) [reply .github/workflows/build_images.yml:50]: Nope

File: .github/workflows/build_images.yml [hunk 2]
```diff
@@ -0,0 +1,61 @@
+name: Build docker images
+
+on:
+  push:
+    branches: [ main ]
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
```
Comment (szaimen) [inline .github/workflows/build_images.yml:9]: Needs to skip if github_branch != nextcloud-releases
See https://docs.github.com/en/actions/learn-github-actions/environment-variables

File: .github/workflows/build_images.yml (hunk 1 above)
Comment (szaimen) [reply .github/workflows/build_images.yml:50]: Yes

File: .github/workflows/build_images.yml (hunk 2 above)
Comment (szaimen) [reply .github/workflows/build_images.yml:9]: not needed anymore

**Comment by szaimen** [comment]
Hi @pestotoast we will build images in this repository using your workflow: https://github.com/nextcloud-releases/all-in-one/pull/3. I hope you don't mind that I am closing this PR because the workflow will only be present on https://github.com/nextcloud-releases/all-in-one
See https://github.com/nextcloud-releases/all-in-one/pull/3

My other changes will be ported to a different branch. Thank your for your patience and the workflow! :)

**Pull request closed without merging**